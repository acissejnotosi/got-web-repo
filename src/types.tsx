export interface House {
  url: string
  name: string
  region: string
  coatOfArms: string
  words: string
  titles: string[]
  seats: string[]
  currentLord: string
  heir: string
  overlord: string
  founded: string
  founder: string
  diedOut: string
  ancestralWeapons: string[]
  cadetBranches: string[]
  swornMembers: string[]
}

export type HouseSigil = {
  url: string
  position: {
    top: string
    left: string
  }
  image: string
}

export type SigilProps = {
  handleSideContainer: (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void
  houseSigil: HouseSigil
  houseName: string
}

export type SideContainerProps = {
  showContainer: boolean
  showHouse: House | undefined
  handleCloseButton: () => void
}

export type HouseComponentType = {
  houseDetails: House | undefined
}

export type DisplayDetailsListType = {
  details: string[] | undefined
  title: string
}

export type DisplayDetailsTextType = {
  detail: string | undefined
  title: string
}
