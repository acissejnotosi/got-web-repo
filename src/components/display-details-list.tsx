import { DisplayDetailsListType } from '../types'

export const DisplayDetailsList = ({
  details,
  title
}: DisplayDetailsListType): JSX.Element => (
  <>
    {details?.toString() !== '' && (
      <>
        <h2>{title}</h2>
        <ul>
          {details?.map(detail => (
            <li key={detail.toString().concat('key')}>{detail}</li>
          ))}
        </ul>
      </>
    )}
  </>
)
