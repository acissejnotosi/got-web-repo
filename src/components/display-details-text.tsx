import { DisplayDetailsTextType } from '../types'

export const DisplayDetailsText = ({
  detail,
  title
}: DisplayDetailsTextType): JSX.Element => (
  <>
    {detail?.toString() !== '' && (
      <>
        <h2>{title}</h2>
        <p>{detail}</p>
      </>
    )}
  </>
)
