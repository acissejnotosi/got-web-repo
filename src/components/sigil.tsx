import { SigilProps } from '../types'

export const Sigil = ({
  handleSideContainer,
  houseSigil,
  houseName
}: SigilProps): JSX.Element => (
  <div
    className='sigil-container'
    style={{
      top: houseSigil?.position.top,
      left: houseSigil?.position.left
    }}
  >
    <button
      aria-label='Sigil button'
      type='button'
      className='button-sigil'
      data-testid='button-sigil'
      value={houseName}
      onClick={e => handleSideContainer(e)}
      style={{
        backgroundImage: `url(${houseSigil?.image})`
      }}
    />
    <span className='house-name' data-testid='house-name'>
      {houseName}
    </span>
  </div>
)
