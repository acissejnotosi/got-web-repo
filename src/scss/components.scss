/* everything that is reusable â€“ buttons, navbars, cards, and so on */

@font-face {
  font-family: "MyWebFont";
  src: url("../resources/fonts/Kameron-Regular.ttf") format("truetype");
}

body {
  margin: 0;
  font-family: "MyWebFont", -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell",
    "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background-color: #0d0d0d;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace;
}

#loader {
  position: fixed;
  left: 50%;
  top: 50%;
  z-index: 101;
}

#error-text {
  padding: 10px;
  color: #f2f2f2;
}

.map-container {
  position: relative;
  img {
    &#got-map {
      top: 0%;
      left: 0%;
      width: max-content;
      height: auto;
    }
  }

  .map-body {
    img {
      &#got-logo {
        position: absolute;
        z-index: 2;
        top: -60px;
        left: 1000px;
      }
    }
  }
}

.sigil-container {
  position: absolute;
  display: flex;
  flex-direction: column;
  align-items: center;
  .button-sigil {
    z-index: 3;
    width: 50px;
    height: 60px;
    background-repeat: no-repeat;
    margin: 0px;
    overflow: hidden;
    border: none;
    cursor: pointer;
    background: round;
    opacity: 0.8;
  }

  .house-name {
    z-index: 3;
    color: #f2f2f2;
    background-color: #0d0d0d;
    opacity: 0.7;
    padding: 5px 10px 5px 10px;
  }
  &:hover {
    .button-sigil {
      opacity: 1;
    }
    .house-name {
      opacity: 1;
    }
  }
}

#side-container {
  position: fixed;
  top: 0;
  right: 0;
  height: 100%;
  width: 400px;
  background: #333333;
  -webkit-overflow-scrolling: touch;
  box-shadow: 0 0 10px 0 rgb(0 0 0 / 70%);
  z-index: 4;
  transition: width 0.1s;
  overflow: hidden;
  .side-content {
    overflow: auto;
    display: flex;
    flex-direction: column;
    padding: 50px 15px;
    height: 100%;
    color: #f2f2f2;
  }
  .side-box {
    box-shadow: 0 2px 6px rgb(50 50 93 / 20%);
    background-color: #454545;
    padding: 10px;
    color: #f2f2f2;
    margin: 0 0 20px 0;
    border-radius: 4px;
  }
  .side-content-main-title {
    text-align: center;
    margin: 0 0 10px 0;
  }
}

.close-btn {
  left: 10px;
  width: 26px;
  height: 26px;
  color: #0d0d0d;
  position: absolute;
  top: 10px;
  opacity: 0.3;
  transition: opacity 0.2s;
  cursor: pointer;
  z-index: 2;
  padding: 3px 2px;
  line-height: 8px;
  font-size: 2em;
  border: none;
  background: transparent;
}

.close-btn:before {
  box-sizing: border-box;
}

.close-btn:before {
  box-sizing: border-box;
}
.close-btn:hover {
  color: #f1f1f1;
}
