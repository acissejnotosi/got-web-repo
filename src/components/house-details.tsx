import { HouseComponentType } from '../types'
import { DisplayDetailsList } from './display-details-list'
import { DisplayDetailsText } from './display-details-text'

export const HouseDetails = ({
  houseDetails
}: HouseComponentType): JSX.Element => (
  <div className='side-content'>
    <h1 className='side-content-main-title'>{houseDetails?.name}</h1>
    <div className='side-box'>
      <DisplayDetailsText detail={houseDetails?.region} title='Region' />
      <DisplayDetailsText
        detail={houseDetails?.coatOfArms}
        title='Coat Of Arms'
      />
      <DisplayDetailsText detail={houseDetails?.words} title='Words' />
      <DisplayDetailsList details={houseDetails?.titles} title='Titles' />
      <DisplayDetailsList details={houseDetails?.seats} title='Seats' />
      <DisplayDetailsText detail={houseDetails?.founded} title='Founded' />
      <DisplayDetailsList
        details={houseDetails?.ancestralWeapons}
        title='Ancestral Weapons'
      />
    </div>
  </div>
)
