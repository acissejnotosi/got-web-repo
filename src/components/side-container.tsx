import { SideContainerProps } from '../types'
import { HouseDetails } from './house-details'

export const SideContainer = ({
  showContainer,
  showHouse,
  handleCloseButton
}: SideContainerProps): JSX.Element => {
  if (showContainer) {
    return (
      <div id='side-container' style={{ width: '500px' }}>
        <button type='button' className='close-btn' onClick={handleCloseButton}>
          &times;
        </button>
        <HouseDetails houseDetails={showHouse} />
      </div>
    )
  }
  return <div id='side-container' style={{ width: '0px' }} />
}
