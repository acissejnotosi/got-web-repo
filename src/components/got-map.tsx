import React, { useEffect, useState } from 'react'
import '../scss/components.scss'
import { useHistory } from 'react-router-dom'
import Loader from 'react-loader-spinner'
import gotMap from '../resources/images/got-map.jpg'
import gotLogo from '../resources/images/got-logo.png'
import { House } from '../types'
import { Sigil } from './sigil'
import { SideContainer } from './side-container'
import { HOUSES } from '../resources/constant'

function GOTMap (): JSX.Element {
  const [houseList, setHouseList] = useState<House[]>()
  const [errorMessage, setErrorMessage] = useState<string>('')
  const [showSideContainer, setShowSideContainer] = useState<boolean>(false)
  const [selectedHouse, setSelectedHouse] = useState<House>()
  const { push } = useHistory()

  const handleSideContainer = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    setShowSideContainer(true)
    setSelectedHouse(
      houseList?.find(house => house.name === e.currentTarget.value)
    )
  }

  const handleCloseSideContainer = () => {
    setShowSideContainer(false)
    setSelectedHouse(undefined)
  }

  const handleError = (error: string) => {
    setErrorMessage(error)
  }

  useEffect(() => {
    fetch('https://www.anapioficeandfire.com/api/houses')
      .then(response => response.json())
      .then(response => setHouseList(response))
      .catch(
        error =>
          error &&
          handleError(
            'It has failed to load Game Of Thrones houses details! Try again later.'
          )
      )
  }, [])

  useEffect(() => {
    if (selectedHouse !== undefined) {
      push(`/${selectedHouse?.name}`)
    } else {
      push(`/`)
    }
  }, [selectedHouse])

  if (houseList) {
    return (
      <div className='map-container'>
        <img id='got-map' src={gotMap} alt='map' />
        <div className='map-body'>
          <img id='got-logo' src={gotLogo} alt='logo' />
          {houseList?.map(house => {
            const houseFound = HOUSES.find(h => h.url === house.url)
            if (houseFound) {
              return (
                <Sigil
                  handleSideContainer={handleSideContainer}
                  houseSigil={houseFound}
                  houseName={house.name}
                />
              )
            }
            return null
          })}
          <SideContainer
            showContainer={showSideContainer}
            showHouse={selectedHouse}
            handleCloseButton={handleCloseSideContainer}
          />
        </div>
      </div>
    )
  }
  if (errorMessage !== '') return <h2 id='error-text'>{errorMessage}</h2>
  return (
    <div id='loader'>
      <Loader type='ThreeDots' color='#F2F2F2' height={100} width={100} />
    </div>
  )
}

export default GOTMap
