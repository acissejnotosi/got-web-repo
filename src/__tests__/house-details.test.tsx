import { expect, test } from '@jest/globals'
import { create } from 'react-test-renderer'
import { HouseDetails } from '../components/house-details'
import { House } from '../types'

test('renders correctly with one house', () => {
  const house: House = {
    url: 'https://anapioficeandfire.com/api/houses/1',
    name: 'House Algood',
    region: 'The Westerlands',
    coatOfArms:
      'A golden wreath, on a blue field with a gold border(Azure, a garland of laurel within a bordure or)',
    words: '',
    titles: [''],
    seats: [''],
    currentLord: '',
    heir: '',
    overlord: 'https://anapioficeandfire.com/api/houses/229',
    founded: '',
    founder: '',
    diedOut: '',
    ancestralWeapons: [''],
    cadetBranches: [],
    swornMembers: []
  }
  const tree = create(<HouseDetails houseDetails={house} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders correctly with list inside house details', () => {
  const house: House = {
    url: 'https://www.anapioficeandfire.com/api/houses/2',
    name: 'House Allyrion of Godsgrace',
    region: 'Dorne',
    coatOfArms: 'Gyronny Gules and Sable, a hand couped Or',
    words: 'No Foe May Pass',
    titles: [''],
    seats: ['Godsgrace'],
    currentLord: 'https://www.anapioficeandfire.com/api/characters/298',
    heir: 'https://www.anapioficeandfire.com/api/characters/1922',
    overlord: 'https://www.anapioficeandfire.com/api/houses/285',
    founded: '',
    founder: '',
    diedOut: '',
    ancestralWeapons: [''],
    cadetBranches: [],
    swornMembers: [
      'https://www.anapioficeandfire.com/api/characters/298',
      'https://www.anapioficeandfire.com/api/characters/1129',
      'https://www.anapioficeandfire.com/api/characters/1301',
      'https://www.anapioficeandfire.com/api/characters/1922'
    ]
  }
  const tree = create(<HouseDetails houseDetails={house} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders correctly with no house details', () => {
  const house: House = {
    url: '',
    name: '',
    region: '',
    coatOfArms: '',
    words: '',
    titles: [''],
    seats: [''],
    currentLord: '',
    heir: '',
    overlord: '',
    founded: '',
    founder: '',
    diedOut: '',
    ancestralWeapons: [''],
    cadetBranches: [],
    swornMembers: []
  }
  const tree = create(<HouseDetails houseDetails={house} />).toJSON()
  expect(tree).toMatchSnapshot()
})
