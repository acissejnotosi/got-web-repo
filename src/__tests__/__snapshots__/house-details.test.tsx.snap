// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with list inside house details 1`] = `
<div
  className="side-content"
>
  <h1
    className="side-content-main-title"
  >
    House Allyrion of Godsgrace
  </h1>
  <div
    className="side-box"
  >
    <h2>
      Region
    </h2>
    <p>
      Dorne
    </p>
    <h2>
      Coat Of Arms
    </h2>
    <p>
      Gyronny Gules and Sable, a hand couped Or
    </p>
    <h2>
      Words
    </h2>
    <p>
      No Foe May Pass
    </p>
    <h2>
      Seats
    </h2>
    <ul>
      <li>
        Godsgrace
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders correctly with no house details 1`] = `
<div
  className="side-content"
>
  <h1
    className="side-content-main-title"
  >
    
  </h1>
  <div
    className="side-box"
  />
</div>
`;

exports[`renders correctly with one house 1`] = `
<div
  className="side-content"
>
  <h1
    className="side-content-main-title"
  >
    House Algood
  </h1>
  <div
    className="side-box"
  >
    <h2>
      Region
    </h2>
    <p>
      The Westerlands
    </p>
    <h2>
      Coat Of Arms
    </h2>
    <p>
      A golden wreath, on a blue field with a gold border(Azure, a garland of laurel within a bordure or)
    </p>
  </div>
</div>
`;
